### ISSUE TITLE
How do you handle error responses properly?

### ISSUE THREAD

**Issue by RavenHursT** [request/request-promise#48]
So I'm using request-promise in a script that I have that loops through a list of urls and fires of requests.  then I want to do something w/ the data received once all the requests are complete.

I have the following:

````javascript
var rp = require('request-promise');

rp.get({
            uri: 'http://httpstat.us/500',
            transform: function(body, res){
                res.data = JSON.parse(body);
                return res;
            }
        }).then(function(res){
            results.push(res.data);
        })
        .catch(function(err){
            should.throw.error.to.console();
            var respErr  = JSON.parse(err.error);
            var errorResult = {
                origUrl: respErr.origUrl,
                error: respErr
            };
            results.push(errorResult);
        });
````

As you can see.. http://httpstat.us/500 throws a 500, which causes the .catch() block to be ran. I'm forcing an error.  ```should.throw.error.to.console();``` should throw an error to the console, but instead, the script just exits silently w/o any error code (```Process finished with exit code 0```).

I'm assuming that you're catching the error from node's ```http``` when a page doesn't come back w/ 2xx code and then passing that back to the catch() callback.  But any subsequent errors then end up failing silently.  How in the world do I handle this so that the rest of my code will still throw errors properly?

**Comment by analog-nico** [comment]
Hey @RavenHursT you do the request correctly. All actual request errors as well as responses with status codes other than 2xx go into the catch handler.

I executed your code and this is the `err` which is passed as an argument:
``` js
{ error: '500 Internal Server Error',
  options: 
   { method: 'GET',
     uri: 'http://httpstat.us/500',
     transform: [Function],
     callback: undefined,
     tunnel: false,
     simple: true,
     resolveWithFullResponse: false },
  response: 
   { _readableState: [Object],
     readable: false,
     domain: null,
     _events: [Object],
     _maxListeners: 10,
     socket: [Object],
     connection: [Object],
     httpVersion: '1.1',
     complete: true,
     headers: [Object],
     trailers: {},
     _pendings: [],
     _pendingIndex: 0,
     url: '',
     method: null,
     statusCode: 500,
     client: [Object],
     _consuming: true,
     _dumped: false,
     httpVersionMajor: 1,
     httpVersionMinor: 1,
     upgrade: false,
     req: [Object],
     pipe: [Function],
     addListener: [Function: addListener],
     on: [Function: addListener],
     pause: [Function],
     resume: [Function],
     read: [Function],
     request: [Object],
     toJSON: [Function: responseToJSON],
     caseless: [Object],
     body: '500 Internal Server Error' },
  statusCode: 500 }
```

If I understand your `should` statement correctly, you expect Request-Promise to print the error to the console by itself. That is not the case.

The crash you observe is actually either caused by the `should` statement or the `JSON.parse` in the next line. I removed the `should` statement and ran your code. On the console I got:
```
Possibly unhandled SyntaxError: Unexpected token I
    at Object.parse (native)
    at /Users/nico/Documents/Projekte/request-tests/issue48.js:15:29

Process finished with exit code 0
```

Bottom line you got trouble with error thrown inside the catch handler which you don't handle anymore.

Does this help you to understand this behavior better?

**Comment by RavenHursT** [comment]
No. The should statement is there to throw a fatal on purpose.. I have further code that runs after the error is passed to the catch.. I noticed my code was dying silently and after an hour or so of trial and error console logging, I found that ANY code with an error in it that runs AFTER a response other than 2xx is received will fail absolutely silently.  But any code before catch will fail with errors in the console.  If all responses come back with 2xx, subsequent errors are printed to console.  This is making it rather hard to debug code that runs after a catch block.

**Comment by analog-nico** [comment]
Try this:

``` js
var rp = require('request-promise');

rp.get({
            uri: 'http://httpstat.us/500',
            transform: function(body, res){
                res.data = JSON.parse(body);
                return res;
            }
        }).then(function(res){
            results.push(res.data);
        })
        .catch(function(err){
            should.throw.error.to.console();
            var respErr  = JSON.parse(err.error);
            var errorResult = {
                origUrl: respErr.origUrl,
                error: respErr
            };
            results.push(errorResult);
        })
        .catch(function(err){
            console.error(err); // This will print any error that was thrown in the previous error handler.
        });
```

Please try to understand why this code does not die silently anymore. For that you should look into promises in general. If you have doubts after all I could give some more clarification.

**Comment by RavenHursT** [comment]
Ah.. thank you for the breadcrumbs.. so if I'm understanding this correctly, .catch is a wrapper for the ```catch``` portion of the try-catch in a Promise implementation?  Hence, by adding a second .catch, I'll be allowing for the exception thrown by a bad request to be caught and sen to the console?

Can I accomplish the same thing doing something like this?

````javascript
var rp = require('request-promise');

rp.get({
            uri: 'http://httpstat.us/500',
            transform: function(body, res){
                res.data = JSON.parse(body);
                return res;
            }
        }).then(function(res){
            results.push(res.data);
        }, function(err){
            should.throw.error.to.console();
            var respErr  = JSON.parse(err.error);
            var errorResult = {
                origUrl: respErr.origUrl,
                error: respErr
            };
            results.push(errorResult);
        })
        .catch(function(err){
            console.error(err); // This will print any error that was thrown in the previous error handler.
        });
```

In other words, does the second callback sent to .then() behave in the same manner as the .catch block?

**Comment by analog-nico** [comment]
Yes, the code snippet is equivalent to the one I posted right above. If we would have no async code but instead everything is sync we would write the same as your last code snippet like this:

``` js
try {

    try {

        var res = rp.get(...);
        results.push(res.data);

    } catch (e) {
        should.throw.error.to.console();
        // Since the should throws an error it is now obvious why we need the second catch.
    }

} catch (e) {
    console.error(err);
}
```

Very important here is that Request-Promise does not throw an error which then would cause the behavior you initially reported in this issue. The error Request-Promise returns in the first error handler can be handled in this error handler alone. You just need a second error handler because your code in the first error handler itself throws another error. To catch this second error you need the second error handler.

Request-Promise uses [Bluebird](https://github.com/petkaantonov/bluebird) as the library which implements the promise functionality. Since Promises conform to a [certain standard](https://promisesaplus.com) there are many other libraries which also implement the same functionality. I suggest you read more about how promises work. There are plenty of tutorial etc. on the web.

Apart from the standard each Promise library handles errors which were NOT properly handled by your code a little different. Here is [how Bluebird handles unhandled errors](https://github.com/petkaantonov/bluebird#error-handling). It explains you why in the beginning you saw `Possibly unhandled SyntaxError` in the console.

I close this issue now because there is no bugfix needed for Request-Promise. However, if you would like to have some more clarification feel free to add another comment to this conversation.

**Comment by RavenHursT** [comment]
Awesome.  Thank you so much for the detailed explaination, @analog-nico.  As for what helped me understand this was this write-up that I found on promises, and the error handling thereof: http://dailyjs.com/2014/02/20/promises-in-detail/

Maybe it'll help someone else in the future.  Thanks again :+1:

**Comment by analog-nico** [comment]
You are welcome. Thanks for the link!

**Comment by viswanadha-kondoju-ck** [comment]
http://dailyjs.com/2014/02/20/promises-in-detail/ is not working

**Comment by jbistis** [comment]
I am using firebase cloud functions and trying to pass the error back to the calling service.
```
 const finalUrl = `${url}/vessels/mmsi/${mmsi}/latest`;
    const options = {
        uri: finalUrl,
        json: true,
        auth: {
            username: config.aquaplot.apiUsername,
            password: config.aquaplot.apiSecret
        }
    }
   
    try {
        const result = await rp(options);
        console.log('result',result);
        return result;
    } catch (error) {
        console.log('we have an error');
        console.log('error',error);
        return error;
    }
```
Function log shows error "error { StatusCodeError: 403 - {"status":"error","message":"Not authorized to access this endpoint"}" which is the correct error I want to send back the calling service.  Tried many approaches...get console error "ERROR Error: Uncaught (in promise): Error: INTERNAL"

Has anyone seen this before?
